src/hotspot/share/gc/shared/referenceProcessor.cpp

```diff
@@ -258,7 +259,11 @@ void DiscoveredListIterator::remove() {
   // Remove Reference object from discovered list. Note that G1 does not need a
   // pre-barrier here because we know the Reference has already been found/marked,
   // that's how it ended up in the discovered list in the first place.
-  RawAccess<>::oop_store(_prev_discovered_addr, new_next);
+  if (_prev_discovered_addr != _refs_list.adr_head()) {
+    _enqueue->enqueue(_prev_discovered_addr, new_next);
+  } else {
+    RawAccess<>::oop_store(_prev_discovered_addr, new_next);
+  }
```

Comment:
What do you think about switching the if-statement around and add a comment explaining why we special case the head?

Reply:
I prefer to keep the common case as first option. Added a comment to it though.